File: vertical/camoufox-cli/src/camoufox_cli/browser.py
"""Browser manager: launches and manages Camoufox instance."""

from __future__ import annotations

import base64
import os

from camoufox.sync_api import Camoufox
from playwright.sync_api import BrowserContext, Page

from .identity import load_or_create, to_launch_kwargs
from .proxy import parse_proxy_settings
from .refs import RefRegistry


_MAX_HISTORY = 200


def _ensure_browser_installed() -> None:
    """Check that the Camoufox browser binary is installed, raise if not."""
    try:
        from camoufox.pkgman import get_path
        get_path("camoufox")
    except Exception:
        raise RuntimeError(
            "Browser not found. Run `camoufox-cli install` to download it."
        )


class TabState:
    """Per-tab state: page pointer, element refs, and navigation history.

    Every named tab shares the single browser context (same fingerprint,
    same cookies/login state) but keeps its own page and view state, so
    concurrent clients don't clobber each other.
    """

    def __init__(self):
        self.page: Page | None = None
        self.refs = RefRegistry()
        # Camoufox spoofs history API for anti-fingerprinting,
        # so we track navigation history ourselves.
        self.history: list[str] = []
        self.history_index: int = -1

    def push_history(self, url: str) -> None:
        # Truncate forward history when navigating to a new page
        self.history = self.history[:self.history_index + 1]
        self.history.append(url)
        # Cap history to avoid unbounded growth in long-lived daemons.
        if len(self.history) > _MAX_HISTORY:
            self.history = self.history[-_MAX_HISTORY:]
        self.history_index = len(self.history) - 1


class BrowserManager:
    def __init__(self, persistent: str | None = None, proxy: str | None = None, geoip: bool = True, locale: str | None = None):
        self._camoufox: Camoufox | None = None
        self._context: BrowserContext | None = None
        self._tabs: dict[str, TabState] = {}
        self._headless: bool = True
        self._persistent = persistent
        self._proxy = proxy
        self._geoip = geoip
        self._locale = locale

    def launch(self, headless: bool = True, tab: str = "default") -> None:
        if self._camoufox is not None:
            return
        self._headless = headless

        _ensure_browser_installed()

        if self._proxy and self._geoip:
            # Geoip resolution lazily downloads the GeoIP db via the
            # rate-limited GitHub API; fetch it through the resilient path first.
            from .install import ensure_mmdb

            ensure_mmdb()

        kwargs: dict = {"headless": headless}
        proxy_settings: dict | None = None
        if self._proxy:
            proxy_settings = parse_proxy_settings(self._proxy)
            kwargs["proxy"] = proxy_settings
            if self._geoip:
                kwargs["geoip"] = True

        if self._persistent:
            # Persistent identity: freeze fingerprint/OS on first launch; reload
            # it on subsequent launches. CLI-passed locale / proxy-derived geo
            # overwrite the stored values so the file tracks current intent.
            os.makedirs(self._persistent, exist_ok=True)
            identity = load_or_create(
                self._persistent,
                locale=self._locale,
                proxy=self._proxy,
                geoip=self._geoip,
            )
            kwargs.update(to_launch_kwargs(identity))
            kwargs["persistent_context"] = True
            kwargs["user_data_dir"] = self._persistent
        elif self._locale:
            # Non-persistent path: locale is a one-shot override, no identity file.
            locales = [s.strip() for s in self._locale.split(",") if s.strip()]
            if locales:
                kwargs["locale"] = locales if len(locales) > 1 else locales[0]

        # A failure partway through leaves _camoufox set but the context/page
        # missing; without this cleanup every later launch() would short-circuit
        # on `_camoufox is not None` and the daemon would be wedged forever.
        try:
            self._camoufox = Camoufox(**kwargs)
            result = self._camoufox.__enter__()

            if self._persistent:
                # persistent_context returns BrowserContext directly
                self._context = result
                pages = self._context.pages
                page = pages[0] if pages else self._context.new_page()
            else:
                # Normal mode: result is Browser. Create an explicit context so
                # more tabs can be added later — the implicit context made by
                # browser.new_page() refuses context.new_page().
                self._context = result.new_context()
                page = self._context.new_page()
            self.state(tab).page = page

            # Workaround: Playwright's Firefox (Juggler) fails proxy auth on HTTPS
            # CONNECT tunnels, raising NS_ERROR_PROXY_AUTHENTICATION_FAILED.
            # Inject Basic auth as an extra HTTP header like WebKit/Chromium do.
            if proxy_settings and proxy_settings.get("username"):
                creds = f"{proxy_settings['username']}:{proxy_settings.get('password', '')}"
                token = base64.b64encode(creds.encode()).decode()
                self._context.set_extra_http_headers(
                    {"Proxy-Authorization": f"Basic {token}"}
                )
        except Exception:
            # Roll back partial state so a later launch() can retry cleanly.
            self.close()
            raise

    def state(self, tab: str) -> TabState:
        """Get (lazily creating) the state record for a named tab."""
        st = self._tabs.get(tab)
        if st is None:
            st = TabState()
            self._tabs[tab] = st
        return st

    def get_page(self, tab: str = "default", create: bool = False) -> Page:
        """Get the tab's page.

        With ``create`` (navigation commands only) a missing/closed page is
        lazily (re)created in the shared context — a new tab gets its own page
        with the same fingerprint and cookies as every other tab. Without it,
        a command on a tab that never opened a page fails loudly instead of
        silently operating on a blank about:blank (which would also leak a
        stray page per misrouted/typo'd tab name).
        """
        ctx = self.get_context()
        st = self.state(tab)
        if st.page is None or st.page.is_closed():
            if not create:
                raise RuntimeError(f"Tab '{tab}' has no open page. Send 'open <url>' first.")
            st.page = ctx.new_page()
        return st.page

    def get_context(self) -> BrowserContext:
        if self._context is None:
            raise RuntimeError("Browser not launched. Send 'open' command first.")
        return self._context

    def get_tabs(self, tab: str = "default") -> list[dict]:
        ctx = self.get_context()
        current = self._tabs.get(tab)
        owners: dict[Page, str] = {}
        for name, st in self._tabs.items():
            if st.page is not None:
                owners.setdefault(st.page, name)
        tabs = []
        for i, p in enumerate(ctx.pages):
            tabs.append({
                "index": i,
                "url": p.url,
                "title": p.title(),
                "active": current is not None and p is current.page,
                "tab": owners.get(p),
            })
        return tabs

    def switch_to_tab(self, tab: str, index: int) -> Page:
        ctx = self.get_context()
        pages = ctx.pages
        if index < 0 or index >= len(pages):
            raise IndexError(f"Tab index {index} out of range (0-{len(pages) - 1})")
        st = self.state(tab)
        st.page = pages[index]
        # The refs and history describe the previous page, not this one — reset
        # them so a stale @ref can't resolve against the newly-adopted page.
        st.refs = RefRegistry()
        st.history = []
        st.history_index = -1
        st.page.bring_to_front()
        return st.page

    def release_tab(self, tab: str) -> None:
        """Close the named tab; last one out shuts down the whole browser.

        This is the 'close' command. Every caller — solo or one of many
        concurrent agents — runs the same thing without knowing about the
        others: it closes only the caller's own page (never a neighbor's,
        which would hijack another agent's tab), and when no live tab
        remains the browser itself is closed. Idempotent: releasing a tab
        that has no open page is a no-op, not an error.
        """
        st = self._tabs.pop(tab, None)
        if st is not None and st.page is not None and not st.page.is_closed():
            try:
                st.page.close()
            except Exception:
                pass
        # Tabs (not context pages) are the refcount: pages nobody owns —
        # window.open popups, pages of misrouted tab names — must not keep
        # the browser alive after the last real tab leaves.
        for other in self._tabs.values():
            if other.page is not None and not other.page.is_closed():
                return
        self.close()

    def go_back(self, tab: str = "default") -> str | None:
        """Go back in the tab's history. Returns the URL or None if at start."""
        st = self.state(tab)
        if st.history_index <= 0:
            return None
        st.history_index -= 1
        url = st.history[st.history_index]
        self.get_page(tab, create=True).goto(url, wait_until="domcontentloaded")
        return url

    def go_forward(self, tab: str = "default") -> str | None:
        """Go forward in the tab's history. Returns the URL or None if at end."""
        st = self.state(tab)
        if st.history_index >= len(st.history) - 1:
            return None
        st.history_index += 1
        url = st.history[st.history_index]
        self.get_page(tab, create=True).goto(url, wait_until="domcontentloaded")
        return url

    def close(self) -> None:
        if self._camoufox is not None:
            try:
                self._camoufox.__exit__(None, None, None)
            except Exception:
                pass
            self._camoufox = None
            self._context = None
            self._tabs.clear()

    @property
    def is_running(self) -> bool:
        return self._camoufox is not None


class TabView:
    """BrowserManager scoped to one named tab.

    Command handlers work against this view, so each client's commands
    route to its own page/refs/history while sharing the browser context
    (fingerprint + cookies) with every other tab.
    """

    def __init__(self, manager: BrowserManager, tab: str):
        self._manager = manager
        self.tab = tab

    @property
    def refs(self) -> RefRegistry:
        return self._manager.state(self.tab).refs

    @property
    def is_running(self) -> bool:
        return self._manager.is_running

    def launch(self, headless: bool = True) -> None:
        self._manager.launch(headless=headless, tab=self.tab)

    def get_page(self, create: bool = False) -> Page:
        return self._manager.get_page(self.tab, create=create)

    def get_context(self) -> BrowserContext:
        return self._manager.get_context()

    def get_tabs(self) -> list[dict]:
        return self._manager.get_tabs(self.tab)

    def switch_to_tab(self, index: int) -> Page:
        return self._manager.switch_to_tab(self.tab, index)

    def push_history(self, url: str) -> None:
        self._manager.state(self.tab).push_history(url)

    def go_back(self) -> str | None:
        return self._manager.go_back(self.tab)

    def go_forward(self) -> str | None:
        return self._manager.go_forward(self.tab)

    def release(self) -> None:
        """Close this tab; the browser shuts down when the last tab leaves."""
        self._manager.release_tab(self.tab)

    def shutdown(self) -> None:
        """Force-close the whole browser regardless of other tabs (close --all)."""
        self._manager.close()

File: vertical/camoufox-cli/js/src/browser.ts
/** Browser manager: launches and manages Camoufox instance. */

import { existsSync, mkdirSync } from "node:fs";
import * as path from "node:path";
import { Camoufox, launchOptions } from "camoufox-js";
import { camoufoxPath } from "camoufox-js/dist/pkgman.js";
import { firefox, type Browser, type BrowserContext, type Page } from "playwright-core";
import { ensureMmdb } from "./install.js";
import { loadOrCreate, toLaunchOptions } from "./identity.js";
import { parseProxySettings } from "./proxy.js";
import { RefRegistry } from "./refs.js";

const MAX_HISTORY = 200;

function ensureBrowserInstalled(): void {
  // Check in-process against the BUNDLED camoufox-js. The old
  // `execFileSync("npx", ["camoufox-js", "path"])` probe was broken three
  // ways: npx resolves camoufox-js from the caller's cwd, so under a global
  // install it downloaded an arbitrary other version from the registry; that
  // needs network at browser-launch time; and in a non-TTY daemon it could
  // hang forever. camoufoxPath(false) has no download side effect — it
  // throws when the browser dir is missing or its version is unsupported.
  try {
    const dir = camoufoxPath(false).toString();
    const launcher =
      process.platform === "darwin"
        ? path.join(dir, "Camoufox.app", "Contents", "MacOS", "camoufox")
        : process.platform === "win32"
          ? path.join(dir, "camoufox.exe")
          : path.join(dir, "camoufox-bin");
    if (!existsSync(launcher)) throw new Error("launcher missing");
  } catch {
    throw new Error(
      "Browser not found. Run `camoufox-cli install` to download it."
    );
  }
}

/**
 * Per-tab state: page pointer, element refs, and navigation history.
 *
 * Every named tab shares the single browser context (same fingerprint,
 * same cookies/login state) but keeps its own page and view state, so
 * concurrent clients don't clobber each other.
 */
export class TabState {
  page: Page | null = null;
  // In-flight ctx.newPage() shared by concurrent commands for this tab, so
  // they don't each create (and orphan) a page. Cleared once it settles.
  pagePromise?: Promise<Page>;
  refs = new RefRegistry();
  // Camoufox spoofs history API for anti-fingerprinting,
  // so we track navigation history ourselves.
  history: string[] = [];
  historyIndex = -1;

  pushHistory(url: string): void {
    this.history = this.history.slice(0, this.historyIndex + 1);
    this.history.push(url);
    if (this.history.length > MAX_HISTORY) {
      this.history = this.history.slice(-MAX_HISTORY);
    }
    this.historyIndex = this.history.length - 1;
  }
}

export class BrowserManager {
  private browser: Browser | null = null;
  private context: BrowserContext | null = null;
  private tabs = new Map<string, TabState>();
  private launching: Promise<void> | null = null;
  private recovering: Promise<void> | null = null;
  // Serializes releaseTab calls; see releaseTab for why.
  private releasing: Promise<void> = Promise.resolve();
  private persistent: string | null;
  private proxy: string | null;
  private geoip: boolean;
  private locale: string | null;

  constructor(persistent: string | null = null, proxy: string | null = null, geoip: boolean = true, locale: string | null = null) {
    this.persistent = persistent;
    this.proxy = proxy;
    this.geoip = geoip;
    this.locale = locale;
  }

  async launch(headless: boolean = true, tab: string = "default"): Promise<void> {
    // Serialize concurrent launches (several tabs' first "open" arriving
    // together) so exactly one browser is created.
    while (this.launching) await this.launching;
    if (this.browser || this.context) return;
    this.launching = this.doLaunch(headless, tab);
    try {
      await this.launching;
    } catch (e) {
      // A failure partway through doLaunch can leave this.browser set but the
      // context/page missing; without this rollback every later launch() would
      // short-circuit on `this.browser || this.context` and wedge the daemon.
      await this.close();
      throw e;
    } finally {
      this.launching = null;
    }
  }

  private async doLaunch(headless: boolean, tab: string): Promise<void> {
    ensureBrowserInstalled();

    if (this.proxy && this.geoip) {
      // Geoip resolution lazily downloads the GeoIP db via the rate-limited
      // GitHub API; fetch it through the resilient path first.
      await ensureMmdb();
    }

    const launchOpts: Record<string, unknown> = { headless };
    let proxySettings: { server: string; username?: string; password?: string } | null = null;
    if (this.proxy) {
      const settings = parseProxySettings(this.proxy);
      proxySettings = settings.proxy;
      launchOpts.proxy = settings.proxy;
      if (this.geoip) {
        launchOpts.geoip = true;
      }
    }

    if (this.persistent) {
      // Persistent identity: freeze fingerprint/OS on first launch; reload
      // it on subsequent launches. CLI-passed locale / proxy-derived geo
      // overwrite the stored values so the file tracks current intent.
      mkdirSync(this.persistent, { recursive: true });
      const identity = await loadOrCreate(
        this.persistent,
        this.locale,
        this.proxy,
        this.geoip,
      );
      Object.assign(launchOpts, toLaunchOptions(identity));
      const opts = await launchOptions(launchOpts);
      this.context = await firefox.launchPersistentContext(this.persistent, opts);
      const pages = this.context.pages();
      this.tabState(tab).page = pages[0] || await this.context.newPage();
    } else {
      if (this.locale) {
        // Non-persistent path: locale is a one-shot override, no identity file.
        const locales = this.locale.split(",").map((s) => s.trim()).filter(Boolean);
        if (locales.length > 0) {
          launchOpts.locale = locales.length > 1 ? locales : locales[0];
        }
      }
      this.browser = await Camoufox(launchOpts) as Browser;
      // Create an explicit context so more tabs can be added later — the
      // implicit context made by browser.newPage() refuses context.newPage().
      this.context = await this.browser.newContext();
      this.tabState(tab).page = await this.context.newPage();
    }

    // Workaround: Playwright's Firefox (Juggler) fails proxy auth on HTTPS
    // CONNECT tunnels, raising NS_ERROR_PROXY_AUTHENTICATION_FAILED.
    // Inject Basic auth as an extra HTTP header like WebKit/Chromium do.
    if (proxySettings?.username) {
      const creds = `${proxySettings.username}:${proxySettings.password ?? ""}`;
      const token = Buffer.from(creds, "utf8").toString("base64");
      await this.context.setExtraHTTPHeaders({
        "Proxy-Authorization": `Basic ${token}`,
      });
    }
  }

  /**
   * Recover after the shared browser/context has died: close what's left and
   * relaunch. Concurrent callers coalesce onto ONE close+relaunch — the daemon
   * handles connections concurrently, so without this two tabs recovering at
   * once would each close+relaunch and one would tear down the browser the
   * other just created. (Python needs no equivalent: its daemon is serial.)
   */
  async recoverDeadBrowser(headless: boolean, tab: string): Promise<void> {
    if (!this.recovering) {
      this.recovering = (async () => {
        await this.close();
        await this.launch(headless, tab);
      })().finally(() => { this.recovering = null; });
    }
    await this.recovering;
  }

  /** Get (lazily creating) the state record for a named tab. */
  tabState(tab: string): TabState {
    let st = this.tabs.get(tab);
    if (!st) {
      st = new TabState();
      this.tabs.set(tab, st);
    }
    return st;
  }

  /**
   * Get the tab's page.
   *
   * With `create` (navigation commands only) a missing/closed page is lazily
   * (re)created in the shared context — a new tab gets its own page with the
   * same fingerprint and cookies as every other tab. Without it, a command on
   * a tab that never opened a page throws instead of silently operating on a
   * blank about:blank (which would also leak a stray page per misrouted tab).
   */
  async getPage(tab: string = "default", create: boolean = false): Promise<Page> {
    const ctx = this.getContext();
    const st = this.tabState(tab);
    if (!st.page || st.page.isClosed()) {
      if (!create) throw new Error(`Tab '${tab}' has no open page. Send 'open <url>' first.`);
      // The daemon handles connections concurrently, so cache the in-flight
      // newPage() — otherwise two commands for the same fresh tab each create
      // a page and one is orphaned in the context forever.
      st.pagePromise ??= ctx.newPage().finally(() => { st.pagePromise = undefined; });
      st.page = await st.pagePromise;
    }
    return st.page;
  }

  getContext(): BrowserContext {
    if (!this.context) throw new Error("Browser not launched. Send 'open' command first.");
    return this.context;
  }

  async getTabsAsync(tab: string = "default"): Promise<{ index: number; url: string; title: string; active: boolean; tab: string | null }[]> {
    const ctx = this.getContext();
    const pages = ctx.pages();
    const current = this.tabs.get(tab);
    const owners = new Map<Page, string>();
    for (const [name, st] of this.tabs) {
      if (st.page && !owners.has(st.page)) owners.set(st.page, name);
    }
    const tabs = [];
    for (let i = 0; i < pages.length; i++) {
      tabs.push({
        index: i,
        url: pages[i].url(),
        title: await pages[i].title(),
        active: current !== undefined && pages[i] === current.page,
        tab: owners.get(pages[i]) ?? null,
      });
    }
    return tabs;
  }

  async switchToTab(tab: string, index: number): Promise<Page> {
    const ctx = this.getContext();
    const pages = ctx.pages();
    if (index < 0 || index >= pages.length) {
      throw new RangeError(`Tab index ${index} out of range (0-${pages.length - 1})`);
    }
    const st = this.tabState(tab);
    st.page = pages[index];
    // The refs and history describe the previous page, not this one — reset
    // them so a stale @ref can't resolve against the newly-adopted page.
    st.refs = new RefRegistry();
    st.history = [];
    st.historyIndex = -1;
    await st.page.bringToFront();
    return st.page;
  }

  /**
   * Close the named tab; last one out shuts down the whole browser.
   *
   * This is the 'close' command. Every caller — solo or one of many
   * concurrent agents — runs the same thing without knowing about the
   * others: it closes only the caller's own page (never a neighbor's,
   * which would hijack another agent's tab), and when no live tab remains
   * the browser itself is closed. Idempotent: releasing a tab that has no
   * open page is a no-op, not an error.
   *
   * Releases are SERIALIZED through a promise chain. Without it, N
   * concurrent closes interleave at the awaits: every one deletes its own
   * map entry first, then every survivor scan sees an emptied map, so all
   * N conclude they are the last tab out — the browser gets torn down N
   * times and isRunning flips false for every in-flight close, which made
   * the server destroy sibling connections before their responses flushed
   * (their tabs were released, but the clients never got the receipt).
   */
  releaseTab(tab: string): Promise<void> {
    const run = this.releasing.then(() => this.doReleaseTab(tab));
    // Keep the chain usable even if a release fails.
    this.releasing = run.catch(() => {});
    return run;
  }

  private async doReleaseTab(tab: string): Promise<void> {
    const st = this.tabs.get(tab);
    this.tabs.delete(tab);
    if (st?.page && !st.page.isClosed()) {
      try { await st.page.close(); } catch {}
    }
    // Tabs (not context pages) are the refcount: pages nobody owns —
    // window.open popups, pages of misrouted tab names — must not keep
    // the browser alive after the last real tab leaves.
    for (const other of this.tabs.values()) {
      if (other.page && !other.page.isClosed()) return;
    }
    await this.close();
  }

  async goBack(tab: string = "default"): Promise<string | null> {
    const st = this.tabState(tab);
    if (st.historyIndex <= 0) return null;
    st.historyIndex--;
    const url = st.history[st.historyIndex];
    await (await this.getPage(tab, true)).goto(url, { waitUntil: "domcontentloaded" });
    return url;
  }

  async goForward(tab: string = "default"): Promise<string | null> {
    const st = this.tabState(tab);
    if (st.historyIndex >= st.history.length - 1) return null;
    st.historyIndex++;
    const url = st.history[st.historyIndex];
    await (await this.getPage(tab, true)).goto(url, { waitUntil: "domcontentloaded" });
    return url;
  }

  async close(): Promise<void> {
    if (this.browser) {
      try { await this.browser.close(); } catch {}
      this.browser = null;
    }
    if (this.context && !this.browser) {
      // persistent context: close context directly
      try { await this.context.close(); } catch {}
    }
    this.context = null;
    this.tabs.clear();
  }

  get isRunning(): boolean {
    return this.browser !== null || this.context !== null;
  }
}

/**
 * BrowserManager scoped to one named tab.
 *
 * Command handlers work against this view, so each client's commands
 * route to its own page/refs/history while sharing the browser context
 * (fingerprint + cookies) with every other tab.
 */
export class TabView {
  constructor(private manager: BrowserManager, readonly tab: string) {}

  get refs(): RefRegistry {
    return this.manager.tabState(this.tab).refs;
  }

  get isRunning(): boolean {
    return this.manager.isRunning;
  }

  launch(headless: boolean = true): Promise<void> {
    return this.manager.launch(headless, this.tab);
  }

  recoverDeadBrowser(headless: boolean = true): Promise<void> {
    return this.manager.recoverDeadBrowser(headless, this.tab);
  }

  getPage(create: boolean = false): Promise<Page> {
    return this.manager.getPage(this.tab, create);
  }

  getContext(): BrowserContext {
    return this.manager.getContext();
  }

  getTabsAsync(): Promise<{ index: number; url: string; title: string; active: boolean; tab: string | null }[]> {
    return this.manager.getTabsAsync(this.tab);
  }

  switchToTab(index: number): Promise<Page> {
    return this.manager.switchToTab(this.tab, index);
  }

  pushHistory(url: string): void {
    this.manager.tabState(this.tab).pushHistory(url);
  }

  goBack(): Promise<string | null> {
    return this.manager.goBack(this.tab);
  }

  goForward(): Promise<string | null> {
    return this.manager.goForward(this.tab);
  }

  /** Close this tab; the browser shuts down when the last tab leaves. */
  release(): Promise<void> {
    return this.manager.releaseTab(this.tab);
  }

  /** Force-close the whole browser regardless of other tabs (close --all). */
  shutdown(): Promise<void> {
    return this.manager.close();
  }
}
